### PR TITLE
Replay Smt states

### PR DESCRIPTION
**Pull request by anwu1219**
Currently in the DNC mode. When solving a subquery, the solver starts from scratch. This PR sets up the mechanism to reuse the tree state from the timed-out parent subquery.

### CODE REVIEW

**Comment by anwu1219** [comment]
Config 1: --dnc --num-workers=8 --initial-divides=3 --restore-tree-states
Config 2: --dnc --num-workers=8 --initial-divides=3

Config 1 (new changes): solve **159** (34496 sec) 
Config 2: solve 151 (25614 sec)

File: src/engine/DnCManager.h [hunk 1]
```diff
@@ -53,7 +53,7 @@ class DnCManager
     /*
       Perform the Divide-and-conquer solving
     */
-    void solve( unsigned timeoutInSeconds );
+    void solve( unsigned timeoutInSeconds, bool restoreTreeStates=true );
```
Comment (guykatzz) [inline src/engine/DnCManager.h:56]: Spacing: restoreTreeStates=true --> restoreTreeStates = true

Also, not crucial, but it's strange that the default configuration is false, but the default parameter to the function is true.

File: src/engine/DnCWorker.cpp [hunk 2]
```diff
@@ -85,9 +88,20 @@ void DnCWorker::popOneSubQueryAndSolve()
...
-        IEngine::ExitCode result = _engine->getExitCode();
+        bool fullSolveNeeded = true; // denotes whether we need to solve the subquery
+        if ( restoreTreeStates && smtState )
+            fullSolveNeeded = _engine->restoreSmtState( *smtState );
+        IEngine::ExitCode result = IEngine::NOT_DONE;
+        if ( fullSolveNeeded )
+        {
+            _engine->solve( timeoutInSeconds );
+            result = _engine->getExitCode();
+        } else
```
Comment (guykatzz) [inline src/engine/DnCWorker.cpp:100]: newline before else

File: src/engine/DnCWorker.h [hunk 3]
```diff
@@ -36,7 +36,7 @@ class DnCWorker
       Pop one subQuery, solve it and handle the result
       Return true if the DnCWorker should continue running
     */
-    void popOneSubQueryAndSolve();
+    void popOneSubQueryAndSolve( bool restoreTreeStates=true );
```
Comment (guykatzz) [inline src/engine/DnCWorker.h:39]: Spacing: restoreTreeStates=true --> restoreTreeStates = true

File: src/engine/IEngine.h [hunk 4]
```diff
@@ -53,6 +55,18 @@ class IEngine
     virtual void restoreState( const EngineState &state ) = 0;
     virtual void setNumPlConstraintsDisabledByValidSplits( unsigned numConstraints ) = 0;
 
+    /*
+      Apply the stack to the newly created SmtCore, returns false if UNSAT is
+      found in this process.
+    */
+
```
Comment (guykatzz) [inline src/engine/IEngine.h:62]: Remove empty line

Comment (guykatzz) [reply src/engine/IEngine.h:62]: Looks like the comments should be swapped?

File: src/engine/SmtCore.cpp [hunk 5]
```diff
@@ -397,6 +397,58 @@ PiecewiseLinearConstraint *SmtCore::chooseViolatedConstraintForFixing( List<Piec
     return candidate;
 }
 
+void SmtCore::replayStackEntry( StackEntry * stackEntry )
```
Comment (guykatzz) [inline src/engine/SmtCore.cpp:400]: remove space after *
Likewise for space after & in the following functions

File: src/engine/StackEntry.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+** Copyright (c) 2017-2019 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved. See the file COPYING in the top-level source
+** directory for licensing information.\endverbatim
+**
+** [[ Add lengthier description here ]]
+
+**/
+
+#ifndef __StackEntry_h__
```
Comment (guykatzz) [inline src/engine/StackEntry.h:16]: Maybe call this an SmtStackEntry?

File: src/engine/StackEntry.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  and also any implied splits that were discovered subsequently.
+*/
+struct StackEntry
+{
+public:
+    PiecewiseLinearCaseSplit _activeSplit;
+    List<PiecewiseLinearCaseSplit> _impliedValidSplits;
+    List<PiecewiseLinearCaseSplit> _alternativeSplits;
+    EngineState *_engineState;
+
```
Comment (guykatzz) [inline src/engine/StackEntry.h:34]: remove empty line

File: src/engine/SmtState.h [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+public:
+    /*
+      Valid splits that were implied by level 0 of the stack.
+    */
+    List<PiecewiseLinearCaseSplit> _impliedValidSplitsAtRoot;
+
+    /*
+      The stack.
+    */
+    List< StackEntry *> _stack;
```
Comment (guykatzz) [inline src/engine/SmtState.h:35]: List< StackEntry *> --> List<StackEntry *>

File: src/engine/SmtState.h [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+#ifndef __SmtState_h__
+#define __SmtState_h__
+
+#include "List.h"
+#include "Map.h"
+#include "PiecewiseLinearConstraint.h"
+#include "StackEntry.h"
+
+class SmtState
```
Comment (guykatzz) [inline src/engine/SmtState.h:24]: Looking at the current SmtCore, it also has a _stateId field. Shouldn't this be part of the SmtState?

File: src/engine/SmtCore.h [hunk 10]
```diff
@@ -115,18 +127,9 @@ class SmtCore
...
-    {
-    public:
-        PiecewiseLinearCaseSplit _activeSplit;
-        List<PiecewiseLinearCaseSplit> _impliedValidSplits;
-        List<PiecewiseLinearCaseSplit> _alternativeSplits;
-        EngineState *_engineState;
-    };
+      duplicate the StackEntry
+    */
+    StackEntry *duplicateStackEntry( const StackEntry &stackEntry );
```
Comment (guykatzz) [inline src/engine/SmtCore.h:132]: Looks like this should be a method of the StackEntry class

File: src/engine/SmtCore.cpp [hunk 11]
```diff
@@ -397,6 +397,58 @@ PiecewiseLinearConstraint *SmtCore::chooseViolatedConstraintForFixing( List<Piec
...
+}
+
+StackEntry * SmtCore::duplicateStackEntry( const StackEntry & stackEntry )
+{
+    StackEntry *copy = new StackEntry();
+
+    copy->_activeSplit = stackEntry._activeSplit;
+    copy->_impliedValidSplits = stackEntry._impliedValidSplits;
+    copy->_alternativeSplits = stackEntry._alternativeSplits;
+    copy->_engineState = NULL;
```
Comment (guykatzz) [inline src/engine/SmtCore.cpp:447]: Why is the engine state not duplicated?

File: src/engine/Engine.cpp [hunk 12]
```diff
@@ -1976,6 +1976,70 @@ void Engine::updateDirections()
...
+            performSymbolicBoundTightening();
+        while ( applyAllValidConstraintCaseSplits() );
+
+        // Step 2: replay the stack
+        for ( auto &stackEntry : smtState._stack )
+        {
+            _smtCore.replayStackEntry( stackEntry );
+            // Do all the bound propagation, and set ReLU constraints to inactive (at
+            // least the one corresponding to the _activeSplit applied above.
+            tightenBoundsOnConstraintMatrix();
```
Comment (guykatzz) [inline src/engine/Engine.cpp:2006]: How did we decide this particular order of tightening actions?

File: src/engine/Engine.cpp [hunk 13]
```diff
@@ -1976,6 +1976,70 @@ void Engine::updateDirections()
...
+                _statistics.print();
+            }
+            _exitCode = Engine::UNSAT;
+            for ( PiecewiseLinearConstraint *p : _plConstraints )
+                p->setActiveConstraint( true );
+            return false;
+        }
+    }
+    for ( PiecewiseLinearConstraint *p : _plConstraints )
+        p->setActiveConstraint( true );
```
Comment (guykatzz) [inline src/engine/Engine.cpp:2034]: This looks like a strange thing to do. Is this in order to satisfy the assertion at the start of solve()?

File: src/engine/DnCManager.h (hunk 1 above)
Comment (anwu1219) [reply src/engine/DnCManager.h:56]: Fixed spacing; set default to false.

File: src/engine/DnCWorker.cpp (hunk 2 above)
Comment (anwu1219) [reply src/engine/DnCWorker.cpp:100]: Done

File: src/engine/DnCWorker.h (hunk 3 above)
Comment (anwu1219) [reply src/engine/DnCWorker.h:39]: Done

File: src/engine/SmtCore.cpp (hunk 5 above)
Comment (anwu1219) [reply src/engine/SmtCore.cpp:400]: Done

File: src/engine/SmtState.h (hunk 8 above)
Comment (anwu1219) [reply src/engine/SmtState.h:35]: Done

File: src/engine/StackEntry.h (hunk 7 above)
Comment (anwu1219) [reply src/engine/StackEntry.h:34]: Done

File: src/engine/IEngine.h (hunk 4 above)
Comment (anwu1219) [reply src/engine/IEngine.h:62]: Yep.. Swapped.

File: src/engine/Engine.cpp (hunk 13 above)
Comment (anwu1219) [reply src/engine/Engine.cpp:2034]: Yes. I'm not sure what is a good solution for this.

File: src/engine/Engine.cpp (hunk 12 above)
Comment (anwu1219) [reply src/engine/Engine.cpp:2006]: It's copied from here:
https://github.com/NeuralNetworkVerification/Marabou/blob/0a5062f4f7933213991c2625590a1a09c21211ab/src/engine/Engine.cpp#L257-L264.

File: src/engine/SmtCore.h (hunk 10 above)
Comment (anwu1219) [reply src/engine/SmtCore.h:132]: Moved to StackEntry

File: src/engine/SmtState.h (hunk 9 above)
Comment (anwu1219) [reply src/engine/SmtState.h:24]: Added the stateId as part of the SmtState

File: src/engine/SmtCore.cpp (hunk 11 above)
Comment (anwu1219) [reply src/engine/SmtCore.cpp]: Discussed offline. Added some clarification.

File: src/engine/Engine.cpp (hunk 13 above)
Comment (anwu1219) [reply src/engine/Engine.cpp]: Remove these two lines. Move the assertion that all PLConstraints are active before solving to the end of preprocessing.

File: src/engine/tests/Test_SmtCore.h [hunk 14]
```diff
@@ -538,26 +538,51 @@ class SmtCoreTestSuite : public CxxTest::TestSuite
...
+        // Examine the second stackEntry
+        stackEntry = *( ++smtState._stack.begin() );
+        ASSERT( stackEntry->_activeSplit == *( ++relu2.getCaseSplits().begin() ) );
+        ASSERT( stackEntry->_alternativeSplits.empty() );
+        ASSERT( stackEntry->_impliedValidSplits.size() == 0 );
+
+        clearSmtState( smtState );
+
+        TS_ASSERT_THROWS_NOTHING( smtCore.popSplit() );
+
```
Comment (guykatzz) [inline src/engine/tests/Test_SmtCore.h:577]: remove empty line

**Review by guykatzz — APPROVED** [review]

**Pull request merged**